src/meta_data.rs

```diff
@@ -172,20 +164,80 @@ impl MetaData {
-        // TODO: select node based on node available space,
-        // volume space requirment, and topology requirment
         let idx = rng.gen_range(0, node_list.len());
-        if let Some(node) = node_list.get(idx) {
-            Ok(node.clone())
+        Ok(node_list
+            .get(idx)
+            .unwrap_or_else(|| {
+                panic!(
+                    "failed to get the {}-th node from returned node list, list={:?}",
+                    idx, node_list,
+                )
+            })
+            .clone())
+    }
+
+    /// Get volume source node ID from request
+    fn get_volume_source_node_id(
+        &self,
+        req: &CreateVolumeRequest,
+    ) -> anyhow::Result<Option<String>> {
+        if !req.has_volume_content_source() {
+            return Ok(None);
+        }
+        let volume_src = req.get_volume_content_source();
+        if volume_src.has_snapshot() {
+            let snapshot_id = &volume_src.get_snapshot().snapshot_id;
+            self.get_snapshot_by_id(snapshot_id)
+                .map(|snapshot| Some(snapshot.node_id))
+                .ok_or(anyhow!(format!(
+                    "failed to get snapshot ID={}",
+                    snapshot_id
+                )))
+        } else if volume_src.has_volume() {
+            let volume_id = &volume_src.get_volume().volume_id;
+            self.get_volume_by_id(volume_id)
+                .map(|volume| Some(volume.node_id))
+                .ok_or(anyhow!(format!("failed to get volume ID={}", volume_id)))
         } else {
-            panic!(
-                "failed to get the {}-th node from returned node list, list={:?}",
-                idx, node_list,
-            );
+            Ok(None)
         }
     }
 
+    /// Select a node to create volume or snapshot
+    pub fn select_node(&self, req: &CreateVolumeRequest) -> anyhow::Result<DatenLordNode> {
+        self.get_volume_source_node_id(req)?.map_or_else(
+            || {
+                debug!("request doesn't have node id, select random node");
+                self.select_random_node()
+            },
+            |node_id| {
+                debug!("select node ID={} from request", node_id);
+
+                if !req.has_accessibility_requirements()
+                    || (req
+                        .get_accessibility_requirements()
+                        .get_requisite()
+                        .iter()
+                        .any(|t| t.get_segments().iter().any(|(_k, v)| v == &node_id))
+                        || req
+                            .get_accessibility_requirements()
+                            .get_preferred()
+                            .iter()
+                            .any(|t| t.get_segments().iter().any(|(_k, v)| v == &node_id)))
+                {
+                    Ok(self
+                        .get_node_by_id(&node_id)
+                        .unwrap_or_else(|| panic!("failed to get node ID={}", node_id)))
```

Comment:
self.get_node_by_id(&node_id).context(format!("failed to get node ID={}", node_id))

no need to use Ok here